routing/DELEGATED_ROUTING_HTTP.md

```diff
@@ -0,0 +1,101 @@
+# ![](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square) Delegated Routing HTTP API
+
+**Author(s)**:
+- Gus Eggert
+
+**Maintainer(s)**:
+
+* * *
+
+**Abstract**
+
+"Delegated routing" is a mechanism for IPFS implementations to use for offloading content routing to another process/server. This spec describes an HTTP API for delegated routing.
+
+# Organization of this document
+
+- [Introduction](#introduction)
+- [Spec](#spec)
+  - [Interaction Pattern](#interaction-pattern)
+  - [Cachability](#cachability)
+  - [Transports](#transports)
+  - [Protocol Message Overview](#protocol-message-overview)
+    - [Known Methods](#known-methods)
+- [Method Upgrade Paths](#method-upgrade-paths)
+- [Implementations](#implementations)
+
+# API Specification
+By default, the Delegated Routing HTTP API uses the `application/json` content type. Clients and servers may optionally negotiate other content types such as `application/cbor`, `application/vnd.ipfs.rpc+dag-json`, etc. using the standard `Accept` and `Content-Type` headers.
+
+- `GET /v1/providers/{CID}`
+    - Reframe equivalent: FindProviders
+    - Response
+        
+        ```json
+        {
+        	"Providers": [
+        		{
+        			"PeerID": "...",
+        			"Multiaddrs": ["...", "..."]
+        			"Protocols": [
+        				{
+        					"Codec": 2320,
+        					"Payload": <opaque data>
+        				}
+        			]
+        		}
+        	]
+        	"NextPageToken": "<token>"
+        }
+        ```
+        
+    - Default limit: 100 providers
+    - Optional query parameters
+        - `transfer` only return providers who support the passed transfer protocols, expressed as a comma-separated list of multicodec IDs such as `2304,2320`,
+        - `transport` only return providers whose published multiaddrs explicitly support the passed transport protocols, such as `/quic` or `/tls/ws`.
+- `GET /v1/providers/hash/{multihash}`
+    - This is the same as `GET /v1/providers/{CID}`, but takes a hashed CID encoded as a multihash
```

Comment:
I like "oblivious" as the name but I don't think this is the place to assert a new name for that entire project.

I don't get what you mean by "makes it easier to write IDL mapping in the future", IDL's I have used can map to arbitrary URL paths just fine.

Reply:
* agree with gus - inventing a new name here seems like it's going to introduce confusion - this isn't quite the same semantics as what 'oblivious' is going to mean in odoh-like contexts.
* propose `/v1/providers-doublehash-sha256/<multihash>` - the same as proposed in the previous reframe spec we want the double hash algorithm specified as part of our strategy for hash rotation if needed.
* `{multihash}` is already a hash, so `doublehash` or some equivalent might be less confusing in differentiating what's going on here.

Reply:
Since we have versioning in URL path, `/hashed-providers/`  feels sufficient? (if we ever need to change hash, we would do `/v2/`, right?) 

Placing hash function name in URL looks really, really awkward, [given we want to standardize things like Multicodec ](https://mailarchive.ietf.org/arch/msg/dispatch/Q9aUoF01Upbvl7STjJvjoU8hlHM/);)

Reply:
We could use multicodec id in the path?  Like `/v1/providers/doublehash/18` for sha256?

(I like calling it "doublehash" to prevent confusion.)

> if we ever need to change hash, we would do /v2/, right?

This seems like an implementation detail? If the API supports specifying the hash fn then servers are free to support whatever hash fns they like, so I don't see why this would necessitate a version bump.

Reply:
I'm fine with either of @lidel or @guseggert 's most recent suggestions for naming. - just noting that versioning of the hash function is going to be painful for double hashing, so we should have a plan - it's not quite the same as saying this is 'delegated routing v2'? so wanted to make sure we were okay bumping the API version as a result of that if we go that way. putting the multicodec id (or short name) as part of the path works too.

Reply:
Can we defer this decision to some later time? There's no impl of this anyway right now, right?

Reply:
we're actively building it already, as is the DHT one. this seems pretty non-controversial, but would prefer you pick one of the two suggested and leave it in here rather than needing another spec PR + debate in the next weeks

Reply:
Let's keep this IPIP limited to cleartext `GET /routing/v1/providers/{CID}`, allowing us to merge when Kubo 0.18 ships with reference implementation.

Privacy-preserving endpoint spec will be added in future IPIP.